imf-currencies.py

```diff
@@ -71,48 +72,63 @@
-        if countries_currencies.get(country['@value']):
-            currency_code = countries_currencies.get(country['@value'])
-        else:
-            currency_code = ''
-        if 'Series' in dataset:
-            exchange_rates_data = dataset['Series']['Obs']
-            if type(exchange_rates_data) != list: continue
-            for row in exchange_rates_data:
-                writer.writerow({
-                    'Date': fix_date(row['@TIME_PERIOD']),
-                    'Rate': row['@OBS_VALUE'],
-                    'Currency': currency_code,
-                    'Frequency': 'M',
-                    'Source': 'IMF',
-                    'Country code': country['@value'],
-                    'Country': country['Description']['#text'],
-                })
-        # https://datahelp.imf.org/knowledgebase/articles/630877-api
-        # IMF API is rate-limited and allows only 10 requests every 5 seconds
-        time.sleep(0.75)
 
+# ## For each country, write out monthly exchange rate data
+@click.command()
+@click.option('--source', default=DEFAULT_SOURCE, help='Data source. Options: ENSE (National Currency per SDR, end of period), ENSA (National Currency per SDR, average of period), ENDE (Domestic currency per target USD, end of period), ENDA (Domestic currency per target USD, average of period).')
+@click.option('--target', default=DEFAULT_TARGET, help='Conversion target, Options: XDR (combined with ENSE/ENSA source), USD (combined with ENDE, ENDA source).')
+def write_monthly_exchange_rates(source, target):
+    """ For each country, write out monthly exchange rate data.
+    Using click to allow optional parameters source and target.
+    """
+    country_url = 'http://dataservices.imf.org/REST/SDMX_JSON.svc/CompactData/IFS/M.{}.{}_XDC_{}_RATE'
+    xdr_url = 'http://dataservices.imf.org/REST/SDMX_JSON.svc/CompactData/IFS/M.US.ESD{}_XDR_USD_RATE'
+    output_file = 'output/imf_exchangerates{}.csv'.format("" if source is DEFAULT_SOURCE and target is DEFAULT_TARGET
+                                                          else "_{}_{}".format(source, target))
+    os.makedirs('output', exist_ok=True)
+    with open(output_file, "w") as output_csv:  # Include format statement to catch the default
+        writer = csv.DictWriter(output_csv, FIELDNAMES)
+        writer.writeheader()
+        for i, country in enumerate(imf_countries):
+            print("Getting data for {}".format(country))
+            # There is a different API URL for XDR
+            if country['@value'] == 'XDR':
+                rc = requests.get(xdr_url.format(source[-1])).json()  # Monthly average/end of period for consistency
+            else:
+                try:
+                    rc = requests.get(country_url.format(country['@value'], source, target)).json()
+                except json.decoder.JSONDecodeError:
+                    time.sleep(2)
+                    rc = requests.get(country_url.format(country['@value'])).json()
```

Comment:
I think this line needs to be updated with the new params, i.e. it should look the same as line 106

Reply:
@andylolz Hi sorry this took a while! For some reason, I was not notified of an update, and was working on other things! Cheers for the review. Updated the line.

Reply:
Nice – thanks!